examples/ChefPolicyfileWorkflow.md

```diff
@@ -0,0 +1,893 @@
+Server and ready to be consumed. What next? Now for each Deployable Policy, we
+repeat the following:
+    ```
+    # cd into root of Policy directory
+    # chef update will pull the latest base-linux policy down from the Chef
+    # Infra Server
+
+    chef update
+    kitchen converge
+
+    # does all look good? then;
+    
+    chef export Policyfile.lock.json ./output -a
+    chef push-archive dev-sandbox ./output/ymir-*-<64 char hash>.tgz
+    kitchen destroy
+    ```
+
+The concludes the _manual_ way of doing things, in the next how-to, let's tie 
+it all together with automation!
+
+* [Guide to PolicyFile Automation](./ChefPolicyFileAutomation.md)
+
+
+## Links
+* **ymir policies:**
+  * [ymir-lb](https://github.com/danielcbright/ymir-lb-policy)
+  * [ymir-api](https://github.com/danielcbright/ymir-api-policy)
+  * [ymir-cache](https://github.com/danielcbright/ymir-cache-policy)
+  * [ymir-backend](https://github.com/danielcbright/ymir-backend-policy)
+* **pipeline files:**
+  * [Jenkins: Create Policy Job](https://github.com/danielcbright/policyfile-create-PFP)
+  * [Jenkins: Create Policy Template](https://github.com/danielcbright/policyfile-template-PFP)
+  * [Jenkins: Policy CI](https://github.com/danielcbright/jenkins-policyfile-lib-PFP)
+  * [Jenkins: Policy CD](https://github.com/danielcbright/policyfile-publish-PFP)
+
+# Appendix
+
+## Attributes
+There are a lot of ways to use attributes with Policyfiles, Michael Hedgpeth has
+written a great blog post about it [here](http://hedge-ops.com/policyfile-attributes/).
+But in the spirit of keeping things simple, and building upon simplicity, we're
+going to cover the two main methods for defining attributes with Policyfiles.
+
+### Method 1: Attribute "Hoisting"
+This is the method that is recommended by Chef and is used throughout the 
+Policyfile Workflow and Automation Guides.
+
+#### Process
+* Create a `Policyfile.rb`, no Policyfile Cookbook is even necessary.
+* Include policies, set runlists, define cookbook sources etc.. just as with the
+workflow described in this document. The only thing is there will be no:
+    ```
+    cookbook 'your-policy-name', path: '.'
+    ```
+    This means that all cookbooks are coming from outside sources.
+* From *inside* the `Policyfile.rb`, you will define any and all attributes that
+are specific to the policy, using the example in Method 1, this is how it would
+look inside a `Policyfile.rb` (why someone would change the MongoDB port, I have
+no idea, but it's just an example :)):
+    ```
+    default['local']['mongodb']['config']['mongod']['net']['port'] = 27017
+    default['dev']['mongodb']['config']['mongod']['net']['port']   = 27018
+    default['qa']['mongodb']['config']['mongod']['net']['port']    = 27019
+    default['prod']['mongodb']['config']['mongod']['net']['port']  = 27020
+    ```
+    In the cookbook, I would reference the above attribute as:
+    ```
+    node['mongodb']['config']['mongod']['net']['port']
+    ```
```

Comment:
Note that you can equivalently write a role/wrapper/whatever cookbook that sets this value in an attributes.rb file based on the value of the policy_group using a case statement or other logic.

The hoist mechanism was added to policyfiles because it wasn't possible to write logic in a policyfile due it being rendered into JSON and uploaded to the server.  Hoisting in attribute files doesn't need any support, so it wasn't engineered and wasn't given a name because you get it for free.  Engineering hosting in policyfiles, however, was never intended to replace using attribute files in wrapper cookbooks.

A lot of policyfile users push attributes down into the cookbooks and are perfectly happy with that.  The problem that comes with that is usually around ACL control over the cookbooks, which then drives people to grant write access the policyfiles where you push attributes up into policyfiles (which does not open up arbitrary code-injection).  Once you really want to leverage that, then the deficiency of policyfiles being just JSON and not code becomes apparent, and hoisting fills the gap for the really common requirement of switching based on policy_group.

The advantage of attributes in attributes files for users that are completely unconcerned about ACL issues is that then you have the full power of ruby at your disposal which is running on the node, so you can dynamically inspect the environment the node is in (so lets say you can migrate to pinging some local service directory to find out information about the datacenter/region that the node is running in, rather than hardcoding attribute values in the policyfiles).  That approach leads to fewer knobs being engineered in policyfiles which leads to less complexity in the description of the architecture.

Reply:
@lamont-granquist thanks for the feedback, this is exactly the type of feedback I need to make this better! I'm going to use some of your wording and update the attributes section to give more information for folks on why they should use hoisting in some cases, and attributes files in others. It seems like both ways have their pros and cons. I'm also thinking of adding on an "advanced" section, where we can get into more details with even better examples down the road, what do you think?